homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,378 @@
+        self.end_datetime: Optional[datetime] = end_datetime
+        self.entity_id: str = entity_id
+        self.recurrence: Optional[rrule] = recurrence
+        self.schedule_id: str = uuid4().hex
+        self.start_datetime: datetime = start_datetime
+
+        self.instance_duration: Optional[timedelta]
+        if self.end_datetime:
+            self.instance_duration = self.end_datetime - self.start_datetime
+        else:
+            self.instance_duration = None
+
+    def __str__(self) -> str:
+        """Define the string representation of this schedule."""
+        return (
+            f'<Schedule start="{self.start_datetime}" '
+            f'end="{self.end_datetime}" rrule="{self.recurrence}">'
+        )
+
+    @property
+    def active(self) -> bool:
+        """Return whether the schedule has an active instance."""
+        # Instances with no end datetimes are one-shots and are not active:
+        if not self.end_datetime:
+            return False
+
+        now = dt_util.utcnow()
+
+        # If we're between the initial start and the end time, we're active:
+        if self.start_datetime <= now <= self.end_datetime:
+            return True
+
+        # If we have no recurrence (and we've already established that we're outside the
+        # initial start and end datetimes), we aren't active:
+        if not self.recurrence:
+            return False
+
+        # If there's no "before" datetime, that means we're at the beginning of the
+        # recurrence; since we've already established that there _is_ a recurrence,
+        # we're active:
+        last_start_dt = self.recurrence.before(now, inc=True)
+        if not last_start_dt:
+            return True
+
+        # Lastly, look at the recurrence just before right now and see
+        # if we're still inside it; if we are, we're active:
+        last_end_dt = last_start_dt + self.instance_duration
+        return last_start_dt <= now <= last_end_dt
+
+    @property
+    def expired(self) -> bool:
+        """Return whether the schedule has expired."""
+        now = dt_util.utcnow()
+
+        # If a recurrence exists and there's at least one instance of it in the future,
+        # we're not expired:
+        if self.recurrence and self.recurrence.after(now, inc=True):
+            return False
+
+        # If either the initial start datetime or the end datetime (if it exists) are in
+        # in the future, we're not expired:
+        if self.start_datetime >= now or (
+            self.end_datetime and self.end_datetime >= now
+        ):
+            return False
+
+        return True
+
+    @property
+    def state(self) -> str:
+        """Return the current state of the schdule."""
+        if self.expired:
+            return STATE_EXPIRED
+        if not self._is_on:
+            return STATE_OFF
+        return STATE_ON
+
+    @property
+    def is_on(self) -> bool:
+        """Return whether the schedule is enabled."""
+        return self._is_on
```

Comment:
```suggestion
        return self._async_trigger_listener is not None
```

Reply:
Check my latest. I think this listener belongs in `ScheduleInstance`; in that case, what about keeping this boolean?